File: examples/python/interceptors/streaming/greeter_server.py
# Copyright 2023 gRPC authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""The Python implementation of the GRPC hellostreamingworld.MultiGreeter server."""

import logging
from concurrent import futures

import grpc

import hellostreamingworld_pb2
import hellostreamingworld_pb2_grpc

def _wrap_handler(handler, behavior_wrapper):
    factory = grpc.unary_stream_rpc_method_handler
    behavior = handler.unary_stream
    return factory(
        behavior_wrapper(behavior),
        request_deserializer=handler.request_deserializer,
        response_serializer=handler.response_serializer,
    )


def _wrap_request_iterator(interceptor_id, request_iterator):
    for i, request in enumerate(request_iterator):
        logging.info(f"{interceptor_id}: before request {i} sent")
        yield request
        logging.info(f"{interceptor_id}: after request {i} sent")


def _wrap_response_iterator(interceptor_id, response_iterator):
    for i, response in enumerate(response_iterator):
        logging.info(f"{interceptor_id}: before response {i} returned")
        yield response
        logging.info(f"{interceptor_id}: after response {i} returned")
    logging.info(f"{interceptor_id}: after RPC")


class ServerInterceptor(grpc.ServerInterceptor):

    def __init__(self, interceptor_id: str):
        self._interceptor_id = interceptor_id

    def intercept_service(self, continuation, handler_call_details):
        def wrap_behavior(behavior):
            def _unary_stream(request, context):
                logging.info(f"{self._interceptor_id}: before RPC")
                response_iterator = behavior(request, context)
                new_response_iterator = _wrap_response_iterator(self._interceptor_id, response_iterator)
                return new_response_iterator

            return _unary_stream

        handler = continuation(handler_call_details)
        return _wrap_handler(handler, wrap_behavior)


class MultiGreeter(hellostreamingworld_pb2_grpc.MultiGreeterServicer):
    def sayHello(self, request, context):
        num_greetings = int(request.num_greetings)
        for i in range(num_greetings):
            logging.info(f"SayHelloServerStreaming: received request {i}")
            yield hellostreamingworld_pb2.HelloReply(message=f"Hello number {i}, {request.name}!")


def serve():
    server = grpc.server(
        futures.ThreadPoolExecutor(max_workers=1),
        interceptors=[
            ServerInterceptor("interceptor1"),
            ServerInterceptor("interceptor2"),
            ServerInterceptor("interceptor3"),
        ],
    )
    hellostreamingworld_pb2_grpc.add_MultiGreeterServicer_to_server(MultiGreeter(), server)
    listen_addr = "[::]:50051"
    server.add_insecure_port(listen_addr)
    server.start()
    server.wait_for_termination()


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    serve()

File: examples/cpp/generic_api/greeter_server.cc
// Copyright 2023 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <grpcpp/grpcpp.h>
#include <grpcpp/health_check_service_interface.h>

#include <iostream>
#include <memory>
#include <string>
#include <unordered_set>

#include "absl/flags/flag.h"
#include "absl/flags/parse.h"
#include "absl/strings/str_format.h"
#include "absl/synchronization/mutex.h"

#ifdef BAZEL_BUILD
#include "examples/protos/helloworld.grpc.pb.h"
#else
#include "helloworld.grpc.pb.h"
#endif

ABSL_FLAG(uint16_t, port, 50051, "Server port for the service");

using grpc::ByteBuffer;
using grpc::CallbackGenericService;
using grpc::CallbackServerContext;
using grpc::GenericCallbackServerContext;
using grpc::ProtoBufferReader;
using grpc::ProtoBufferWriter;
using grpc::Server;
using grpc::ServerBuilder;
using grpc::ServerGenericBidiReactor;
using grpc::Status;
using grpc::StatusCode;
using helloworld::HelloReply;
using helloworld::HelloRequest;

// Logic and data behind the server's behavior.
class GreeterServiceImpl final : public CallbackGenericService {
  ServerGenericBidiReactor* CreateReactor(
      GenericCallbackServerContext* context) override {
    if (context->method() == "/helloworld.Greeter/SayHello") {
      // Let the SayHello reactor handle this now on.
      return new SayHelloReactor();
    } else {
      // Forward this to the implementation of the base class returning
      // UNIMPLEMENTED.
      return CallbackGenericService::CreateReactor(context);
    }
  }

  class SayHelloReactor : public ServerGenericBidiReactor {
   public:
    SayHelloReactor() { StartRead(&request_); }

   private:
    Status OnSayHello(const HelloRequest& request, HelloReply* reply) {
      if (request.name() == "") {
        return Status(StatusCode::INVALID_ARGUMENT, "name is not specified");
      }
      reply->set_message(absl::StrFormat("Hello %s", request.name()));
      return Status::OK;
    }

    void OnDone() override { delete this; }
    void OnReadDone(bool ok) override {
      if (!ok) {
        return;
      }
      Status result;
      // Deserialize a request message
      HelloRequest request;
      result = grpc::GenericDeserialize<ProtoBufferReader, HelloRequest>(
          &request_, &request);
      if (!result.ok()) {
        Finish(result);
        return;
      }
      // Call the SayHello handler
      HelloReply reply;
      result = OnSayHello(request, &reply);
      if (!result.ok()) {
        Finish(result);
        return;
      }
      // Serialize a reply message
      bool own_buffer;
      result = grpc::GenericSerialize<ProtoBufferWriter, HelloReply>(
          reply, &response_, &own_buffer);
      if (!result.ok()) {
        Finish(result);
        return;
      }
      StartWrite(&response_);
    }
    void OnWriteDone(bool ok) override {
      Finish(ok ? Status::OK
                : Status(StatusCode::UNKNOWN, "Unexpected failure"));
    }
    ByteBuffer request_;
    ByteBuffer response_;
  };

 private:
  absl::Mutex mu_;
};

void RunServer(uint16_t port) {
  std::string server_address = absl::StrFormat("0.0.0.0:%d", port);
  GreeterServiceImpl service;
  grpc::EnableDefaultHealthCheckService(true);
  ServerBuilder builder;
  // Listen on the given address without any authentication mechanism.
  builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
  // Register "service" as the instance through which we'll communicate with
  // clients. In this case it corresponds to an *synchronous* service.
  builder.RegisterCallbackGenericService(&service);
  // Finally assemble the server.
  std::unique_ptr<Server> server(builder.BuildAndStart());
  std::cout << "Server listening on " << server_address << std::endl;

  // Wait for the server to shutdown. Note that some other thread must be
  // responsible for shutting down the server for this call to ever return.
  server->Wait();
}

int main(int argc, char** argv) {
  absl::ParseCommandLine(argc, argv);
  RunServer(absl::GetFlag(FLAGS_port));
  return 0;
}
